core/src/handshake/manager/internal_engine.rs

```diff
@@ -0,0 +1,270 @@
+//! Defines logic for running the internal matching engine on a given order
+
+use circuits::types::{order::Order, r#match::MatchResult};
+use itertools::Itertools;
+use rand::{seq::SliceRandom, thread_rng};
+use tracing::log;
+
+use crate::{
+    handshake::{
+        error::HandshakeManagerError,
+        manager::{ERR_NO_PRICE_DATA, ERR_NO_WALLET},
+    },
+    state::{wallet::Wallet, NetworkOrder, OrderIdentifier},
+};
+
+use super::{HandshakeExecutor, ERR_NO_ORDER};
+
+// ------------------------
+// | Matching Engine Impl |
+// ------------------------
+
+impl HandshakeExecutor {
+    /// Run the internal matching engine on the given order
+    ///
+    /// TODO: This could be optimized by indexing orders by asset pairs in the
+    /// global state, but this would require further denormalizing the order book
+    /// index. We will hold off on this optimization until we either:
+    ///     1. Determine this code path to be a bottleneck
+    ///     2. Have a better state management abstraction that makes denormalization easier
+    pub(super) async fn run_internal_matching_engine(
+        &self,
+        order: OrderIdentifier,
+    ) -> Result<(), HandshakeManagerError> {
+        log::info!("Running internal matching engine on order {order}");
+        // Lookup the order and its wallet
+        let (network_order, wallet) = self.fetch_order_and_wallet(&order).await?;
+        let my_order = wallet
+            .orders
+            .get(&network_order.id)
+            .ok_or_else(|| HandshakeManagerError::StateNotFound(ERR_NO_ORDER.to_string()))?;
+        let other_orders = self
+            .global_state
+            .read_order_book()
+            .await
+            .get_local_scheduleable_orders()
+            .await;
+
+        // Sample a price to match the order at
+        let (base, quote) = self.token_pair_for_order(my_order);
+        let price = self
+            .fetch_price_vector()
+            .await?
+            .find_pair(&base, &quote)
+            .ok_or_else(|| HandshakeManagerError::NoPriceData(ERR_NO_PRICE_DATA.to_string()))?
+            .2; /* (base, quote, price) */
+
+        // Shuffle the ordering of the other orders for fairness
```

Comment:
Since we can see the full contents of other locally managed orders, should we maybe order by order volume? So that we can minimize the number of fills?

Reply:
It's slightly a pain to do this because of the way the state is setup. It should probably be done eventually I agree, but I'm planning on refactoring the state access at some point as well to support more intricate access patterns like the one you describe. For now, I'd prefer to keep it simple. We will also have a `min_fill_size` field on each order at some point which will aid this issue.